ACI-CNI 5.2.3.2 upgrade

siva-muni opened the pull request:
- Add ACI template
- Add ACI CNI template and images.

snasovich commented on go.mod:
```diff
@@ -4,13 +4,14 @@ go 1.16
 replace (
 	github.com/knative/pkg => github.com/rancher/pkg v0.0.0-20190514055449-b30ab9de040e
-	k8s.io/client-go => k8s.io/client-go v0.21.0
+	github.com/rancher/rke => github.com/noironetworks/rke v0.0.0-20220624061930-035bf38895c6
```
Is this a temporary change until RKE changes in https://github.com/rancher/rke/pull/2963/files are merged?

snasovich commented on go.mod:
```diff
@@ -4,13 +4,14 @@ go 1.16
 	github.com/blang/semver v3.5.1+incompatible
-	github.com/rancher/rke v1.3.0-rc8.0.20210706205346-22b82828ffa0
-	github.com/sirupsen/logrus v1.7.0
+	github.com/rancher/rke v0.0.0-20220624061930-035bf38895c6
```
Is this a temporary change until RKE changes in https://github.com/rancher/rke/pull/2963/files are merged?

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -6596,16 +6596,16 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
-			AciControllerContainer:    "noiro/aci-containers-controller:5.1.1.0.1ae238a",
-			AciGbpServerContainer:     "noiro/gbp-server:5.1.1.0.1ae238a",
-			AciOpflexServerContainer:  "noiro/opflex-server:5.1.1.0.1ae238a",
+			AciCniDeployContainer:     "noiro/cnideploy:5.2.3.2",
```
Updating existing already released versions is not supported. A new version (like `v1.20.15-rancher1-5`) needs to be introduced with bumped ACI versions.

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -6596,16 +6596,16 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
+			AciOpflexServerContainer:  "noiro/opflex-server:5.2.3.2",
 			PodInfraContainer:         "rancher/mirrored-pause:3.6",
-			Ingress:                   "rancher/nginx-ingress-controller:nginx-1.2.1-rancher1",
+			Ingress:                   "rancher/nginx-ingress-controller:nginx-1.2.0-rancher1",
```
What is the reason for Nginx controller version downgrade?

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -7142,6 +7142,92 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
 			WindowsPodInfraContainer:  "rancher/mirrored-pause:3.6",
 			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.18.0",
 		},
+		"v1.21.12-rancher1-2": {
```
Since `v1.21.13-rancher1-1` based on k8s 1.21.13 is already released, we would generally not introduce `v1.21.12-rancher1-2` based on not the latest patch. Is there a reason this updated ACI can't be based on 1.21.13?

In fact, for the upcoming release we're introducing k8s June patch versions - https://github.com/rancher/kontainer-driver-metadata/pull/921. So realistically we would want to introduce updated ACI images in the versions to be added per that PR: `1.23.8-rancher1-1`, `1.22.11-rancher1-1`, `1.21.14-rancher1-1`. The best option would be to wait for the linked PR to be merged and then rebase your PR and make updates to `1.23.8-rancher1-1`, `1.22.11-rancher1-1`, `1.21.14-rancher1-1`. In this case making updates to already existing versions is an acceptable process as these versions were not released yet (not merged to `release-v2.6` branch).

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -7142,6 +7142,92 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
+			AciGbpServerContainer:     "noiro/gbp-server:5.2.3.2",
+			AciOpflexServerContainer:  "noiro/opflex-server:5.2.3.2",
+			PodInfraContainer:         "rancher/mirrored-pause:3.6",
+			Ingress:                   "rancher/nginx-ingress-controller:nginx-1.2.0-rancher1",
```
Same as one of the previous comments - looks like old version of Ingress Controller is referenced. I did not check other images yet, please make sure you're not inadvertently downgrading them.

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -7142,6 +7142,92 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
+			WindowsPodInfraContainer:  "rancher/mirrored-pause:3.6",
+			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.18.0",
+		},
+		"v1.22.9-rancher1-2": {
```
New 1.22 version needs to be defined further down after the latest already defined 1.22 - we want them to be sorted.

snasovich commented on rke/templates/aci.go:
```diff
@@ -1,6 +1,52 @@
 package templates
 
 const AciTemplateV500 = `
+apiVersion: apiextensions.k8s.io/v1
```
Updating the existing template is not supported as will it be used by older versions of ACI that do not support it. You can see an example of adding new template (for Ingress NGINX controller) in this PR https://github.com/rancher/kontainer-driver-metadata/pull/889 and follow the same approach.

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -7142,6 +7142,92 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
+			CoreDNSAutoscaler:         "rancher/mirrored-cluster-proportional-autoscaler:1.8.5",
+			WindowsPodInfraContainer:  "rancher/mirrored-pause:3.6",
+			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.21.1",
+		},
```
Is there a plan to update ACI version for 1.23 version too? Or it's not supported yet?

snasovich commented on go.mod:
```diff
@@ -4,13 +4,14 @@ go 1.16
-	github.com/rancher/rke v1.3.0-rc8.0.20210706205346-22b82828ffa0
-	github.com/sirupsen/logrus v1.7.0
+	github.com/rancher/rke v0.0.0-20220624061930-035bf38895c6
+	github.com/sirupsen/logrus v1.8.1
```
Is there a specific need to bump `logrus` here?

snasovich reviewed (changes requested):
@siva-muni, thank you for submitting this PR.

I've left a few specific comments to address.
Additionally, here are couple general comments:
* The build is failing, please look into why and address: https://github.com/rancher/kontainer-driver-metadata/pull/889/commits
* To merge this PR, we want it to have 2 commits - actual changes and `go generate` (so updates to `data.json`) separately.

siva-muni commented:
[snasovich](https://github.com/snasovich), Thanks for the review, we will address your comments.

siva-muni commented:
We addressed the comments in [91eb1fe](https://github.com/rancher/kontainer-driver-metadata/pull/922/commits/91eb1fef675132f4c1205da6c58b063f44e9759a), [07b240a](https://github.com/rancher/kontainer-driver-metadata/pull/922/commits/07b240a1fcf1f583c2baa6598e05d080a2c779bf), but still we need to test this code once it works fine, We will update, we are working on this.

siva-muni commented:
With these changes, we are seeing the issues, We will fix and update

siva-muni commented:
Made few more changes, testing in progress.

snasovich commented on rke/templates/aci.go:
```diff
@@ -1347,3 +1347,2579 @@ status:
   ipaddr: 0.0.0.0/0
 {{- end}}
 `
+const AciTemplateV523 = `
```
Please put this new template in a separate file to follow the same pattern as used by other templates that have multiple versions (e.g. nginx, see files in https://github.com/rancher/kontainer-driver-metadata/tree/dev-v2.6/rke/templates).

siva-muni replied:
Added separate file for new template in [7e7eabb](https://github.com/rancher/kontainer-driver-metadata/pull/922/commits/7e7eabbba1b4688c0359d90daf1072e09a1fe64f)

snasovich commented on rke/templates/templates.go:
```diff
@@ -167,7 +168,8 @@ func LoadK8sVersionedTemplates() map[string]map[string]string {
 		kdm.Aci: {
-			">=1.17.0-alpha": aciv500,
+			">=1.17.0-alpha <1.20.0-rancher0": aciv500,
+			">=1.20.0-rancher0 <1.23.0-rancher1-1": aciv523,
```
A lot more conditions need to be introduced here to ensure that updated `aciv523` template is used for newly introduced versions only. With the current condition (`>=1.20.0-rancher0 <1.23.0-rancher1-1`) it will be used for any k8s version starting with 1.20, but previously existing versions (e.g. `v1.20.15-rancher1-4`) reference 5.1.1.0 images and thus cannot use the new template that is built for 5.2.3.2.
Please see below an example for `kdm.NginxIngress` that demonstrates more granular conditions to ensure only newly added versions use new template.

siva-muni replied:
Updated required conditions in [7e7eabb](https://github.com/rancher/kontainer-driver-metadata/pull/922/commits/7e7eabbba1b4688c0359d90daf1072e09a1fe64f)

snasovich commented on rke/k8s_rke_system_images.go:
```diff
@@ -7142,6 +7185,94 @@ func loadK8sRKESystemImages() map[string]v3.RKESystemImages {
+			Nodelocal:                 "rancher/mirrored-k8s-dns-node-cache:1.21.1",
+		},
+		// Enabled out of band post v2.6.5
+		"v1.21.13-rancher1-2": {
```
This is out of order with 1.21 version following 1.22. This will be a moot point once we rebase (see overall review comment).

siva-muni replied:
removed this change

snasovich reviewed (changes requested):
I see most of my previous comments are addressed, thank you.

Added some more comments to address.

Also, note that per https://github.com/rancher/kontainer-driver-metadata/pull/922#discussion_r906095796 the best option would be to rebase after PR #921 is merged (should be very soon) to update versions it added with new ACI images /templates instead of introducing new ones. It should be merged very soon, likely today.

snasovich commented:
@siva-muni , #921 is now merged so please go ahead with rebasing this PR and updating versions added by #921 instead of introducing new ones.

siva-muni commented:
[snasovich](https://github.com/snasovich), thanks for the review, I will address the comments

siva-muni commented:
Addressed the comments, We have updated the images in the  versions added by https://github.com/rancher/kontainer-driver-metadata/pull/921 
 - 1.22.11-rancher1-1	
- 1.21.14-rancher1-1	
- 1.20.15-rancher2-1

siva-muni commented:
With these changes, testing is in-progress

snasovich reviewed (dismissed):
@siva-muni , changes LGTM, thank you for addressing my comments.

Please report the results of your testing and once you confirm it's successful and we have one more approval, this can be merged.

cc: @kinarashah

siva-muni commented:
[snasovich](https://github.com/snasovich), We are finished our testing, it was successful.  We are not seeing any issues.

kinarashah commented:
@siva-muni Is there any reason we don't want the images to be updated for 1.23.x and 1.24.x? If users upgrade from 1.22.x to 1.23.x aci version would downgrade, is that expected? Apologies for missing necessary context if it's obvious.

siva-muni committed:
added upgrade support for ACI-CNI v5.2.3.2

siva-muni committed:
go generate

siva-muni commented:
[kinarashah](https://github.com/kinarashah),  we added the new images now for 1.23 and 1.24 versions as well. This is an oversight from our side as we are doing this for the first time.

snasovich commented:
@siva-muni , thank you for following up on this. Are we waiting for additional testing or good to merge once @kinarashah approves?

siva-muni commented:
[snasovich](https://github.com/snasovich), We are done with testing. We are good to merge.

snasovich reviewed (approved):
@siva-muni , thank you for updating the PR and reporting that the testing is successful. LGTM.
@kinarashah , could you please take a peek at it as well and if it looks good to you, please merge it.
